Composer/packages/adaptive-form/src/components/expressions/ExpressionsListMenu.tsx

```diff
@@ -1,62 +1,96 @@
+import React from 'react';
+import { Callout } from 'office-ui-fabric-react/lib/Callout';
+import { FieldToolbar } from '@bfc/code-editor';
+import { useShellApi } from '@bfc/extension-client';
 
 type ExpressionsListMenuProps = {
-  onExpressionSelected: (expression: string) => void;
-  onMenuMount: (menuContainerElms: HTMLDivElement[]) => void;
+  container: HTMLDivElement | null;
+  target: HTMLInputElement | HTMLTextAreaElement | null;
+  clearTarget: () => void;
+  value?: string;
+  onChange: (expression: string) => void;
 };
+
 export const ExpressionsListMenu = (props: ExpressionsListMenuProps) => {
-  const { onExpressionSelected, onMenuMount } = props;
+  const { clearTarget, container, target, value = '', onChange } = props;
+  const { projectId, shellApi } = useShellApi();
 
-  const containerRef = React.createRef<HTMLDivElement>();
+  const [memoryVariables, setMemoryVariables] = React.useState<string[] | undefined>();
 
-  const onExpressionKeySelected = useCallback(
-    (key) => {
-      const insertText = getBuiltInFunctionInsertText(key);
-      onExpressionSelected('= ' + insertText);
-    },
-    [onExpressionSelected]
-  );
+  React.useEffect(() => {
+    const abortController = new AbortController();
+    (async () => {
+      try {
+        const variables = await shellApi.getMemoryVariables(projectId, { signal: abortController.signal });
+        setMemoryVariables(variables);
+      } catch (e) {
+        // error can be due to abort
+      }
+    })();
+  }, [projectId]);
 
-  const onLayerMounted = useCallback(() => {
-    const elms = document.querySelectorAll<HTMLDivElement>('.ms-ContextualMenu-Callout');
-    onMenuMount(Array.prototype.slice.call(elms));
-  }, [onMenuMount]);
-
-  const menuItems = useMemo(
-    () =>
-      expressionGroupingsToMenuItems(
-        builtInFunctionsGrouping,
-        onExpressionKeySelected,
-        onLayerMounted,
-        componentMaxHeight
-      ),
-    [onExpressionKeySelected, onLayerMounted]
-  );
+  React.useEffect(() => {
+    const keyDownHandler = (e: KeyboardEvent) => {
+      if (e.key === 'Escape') {
+        clearTarget();
+      }
+    };
 
-  return (
-    <div ref={containerRef}>
-      <ContextualMenu
-        calloutProps={{
-          onLayerMounted: onLayerMounted,
-        }}
-        directionalHint={DirectionalHint.bottomLeftEdge}
-        hidden={false}
-        items={menuItems}
-        shouldFocusOnMount={false}
-        styles={{
-          container: { maxHeight: componentMaxHeight },
-        }}
-        target={containerRef}
-      />
-    </div>
+    const focusHandler = (e: FocusEvent) => {
+      if (container?.contains(e.target as Node)) {
+        return;
+      }
+
+      if (
+        !e
+          .composedPath()
+          .filter((n) => n instanceof Element)
+          .map((n) => (n as Element).className)
+          .some((c) => c.indexOf('js-lg-toolbar-menu') !== -1)
```

Comment:
change this to js-field-toolbar-menu (in constants and then here)

Reply:
Lol already on my list